### ISSUE TITLE
Cannot call System.Text.Json.Serialization.Converters namespace methods in Unity IL2CPP

### ISSUE THREAD

**Issue by ponahoum** [dotnet/runtime#49772]
Hi everyone,

I'm trying to use the brand new System.Text.Json (using netstandard2.0 5.0.1 build) library with Il2CPP builds in Unity.

When trying to serialize a simple float array float[3] with the library in Unity, I get the error :

`
ExecutionEngineException: Attempting to call method 'System.Text.Json.Serialization.Converters.ArrayConverter'2[[System.Single[], mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.Single, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]::.ctor' for which no ahead of time (AOT) code was generated.
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00000] in <00000000000000000000000000000000>:0 
  at System.RuntimeType.CreateInstanceMono (System.Boolean nonPublic) [0x00000] in <00000000000000000000000000000000>:0 ...
`

It seems that the generic System.Text.Json.Serialization.Converters.ArrayConverter class of System.Text.Json could not be included in the build, and in general all the System.Text.Json.Serialization.Converters* namespace methods.

I had this issue with string and base types before, but I solved it by adding a link.xml to my assets folder to prevent some stripping by the compiler, which for some reason doesn't work for the array converters and other more complex types converters under the namespace  **System.Text.Json.Serialization.Converters**
`

<linker>

    <assembly fullname="System.Text.Json">

        <namespace fullname="System.Text.Json" preserve="all" />

        <namespace fullname="ystem.Text.Json.Serialization" preserve="all" />

        <namespace fullname="System.Text.Json.Serialization.Converters" preserve="all" />

    </assembly>

</linker>

`

Did anyone manage to make System.Text.Json work with Unity Il2CPP ?

**Comment by MichalStrehovsky** [comment]
I would recommend checking with Unity if there's a way to force specific generic instantiation to be compiled. The problem here is with MakeGenericType/MakeGenericMethod-kind of calls - they potentially require new native code to be produced but IL2CPP obviously can't produce new code at runtime.

Utf8Json in connection with [pre code generation](https://github.com/neuecc/Utf8Json/#pre-code-generationunityxamarin-supports) works with Unity.

System.Text.Json is also working on pre code generation for .NET 6 (https://github.com/dotnet/runtime/issues/45441), but it will be done with Roslyn source generators and I'm not sure how that will look like in Unity.

**Comment by ponahoum** [comment]
Thanks @MichalStrehovsky . I actually progressed a little since last night. After reading a lot and seeing @JoshPeterson from Unity post, I understood that unity il2cpp specifically unsupports generic functions with **value types** that are not pre-generated.

So basically Unity IL2CPP serialization with System.Text.Json would work with anything that is a reference type (string[], any object) but not with value types (enum, struct, float[], List<float>) because it needs to call his generic converters such as ArrayConverter, and for example, ArrayConverter<float[]> can not be generated as it takes a value type as an argument, although ArrayConverter<string[]> is acceptable.



To solve this, I found three ways:

- Modify the Dotnet library to expose all converters as public, and then making a UsedOnlyForAOTCodeGeneration() function to generate the generic converters that I need (as described here : [ Unity Scripting Rectrictions](https://docs.unity3d.com/Manual/ScriptingRestrictions.html)). This solution is pretty long and unsafe as it requires recompiling the dotnet library.
-  Unity introduced a file that allows setting generic force to compile for some types, but this is not officially supported (**il2cpp_extra_types.txt**), and I couldn't make it work, it's pretty hard to get the syntax
- The solution I have chosen: make custom converters that make a wrapper around all value types to make them reference types. This is a bit heavy, but the only serious workaround I found for now.

Unity has been working for two years now on supporting value-types generic virtual classes in Il2CPP, and it seems it's not a priority.

**Comment by joshpeterson** [comment]
> Unity has been working for two years now on supporting value-types generic virtual classes in Il2CPP, and it seems it's not a priority.

We've just released value-type generic sharing as an option in Unity 2021.2 alpha. See this post for details:

https://forum.unity.com/threads/il2cpp-build-time-improvements-seeking-feedback.1064135

**Comment by ponahoum** [comment]
Wow @joshpeterson this is great! As it is an alpha I'll probably choose the wrapper option for stability, but will sure migrate to this when it's officially released!

**Comment by MichalStrehovsky** [comment]
Great to hear you're unblocked. Seems like there's nothing else to do - I don't think we would want to expose implementation details of S.T.Json for `UsedOnlyForAOTCodeGeneration` sake - closing if you agree.

**Comment by bdominguez** [comment]
@ponahoum can you give me an example about the third option (your chosen option) so I can better understand it? Thanks!

**Comment by ponahoum** [comment]
Sure !

Let say you want to serialize an instance this object :

`    public class TestObject
    {
        public string aRandomString { get; set; }
        public TestEnum aRandomEnum { get; set; }

        public float[] myFloatArray { get; set; } = new float[] { 1, 2, 3 };
        public string[] strfArr { get; set; } = new string[3] { "abc", "def", "ghi" };
        public enum TestEnum
        {
            EnumFirstTest,
            EnumSecondTest
        }
    }
`
It won't work out of the box with System.Text.Json, because we use float[] and an enum.

So here I basically wrote some converters to handle the enum and the float array:

`
  public class ValueTypeWrapper<T>
    {
        public ValueTypeWrapper(T theValue) {
            this.theValue = theValue;
        }

        public T theValue { get; set; }
    }

    public class ValueTypeConverter<T> : JsonConverter<ValueTypeWrapper<T>>
    {
        public override ValueTypeWrapper<T> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
            if(JsonDocument.TryParseValue(ref reader, out JsonDocument doc)){
                T value = JsonSerializer.Deserialize<T>(doc.RootElement.GetRawText(), options);
                ValueTypeWrapper<T> wrapper = new ValueTypeWrapper<T>(value);
                return wrapper;
            }

            throw new JsonException("Could not parse the value type object");
        }

        public override void Write(Utf8JsonWriter writer, ValueTypeWrapper<T> valueWrapper, JsonSerializerOptions options)
        {
            JsonSerializer.Serialize(writer, valueWrapper.theValue, options);
        }
    }

    public class ArrayConverter<T> : JsonConverter<T[]>
    {
        public override T[] Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
            if (JsonDocument.TryParseValue(ref reader, out JsonDocument doc))
            {
                ValueTypeWrapper<T>[] arr = JsonSerializer.Deserialize<ValueTypeWrapper<T>[]>(doc.RootElement.GetRawText(), options);

                //Get as array
                T[] realArr = new T[arr.Length];
                for (int i = 0; i < realArr.Length; i++)
                {
                    realArr[i] = arr[i].theValue;
                }

                return realArr;
            }

            throw new JsonException("Could not parse the value type object");
        }

        public override void Write(Utf8JsonWriter writer, T[] arr, JsonSerializerOptions options)
        {
            //Convert the array to a reference type
            ValueTypeWrapper<T>[] valueArr = new ValueTypeWrapper<T>[arr.Length];
            for (int i = 0; i < valueArr.Length; i++)
                valueArr[i] = new ValueTypeWrapper<T>(arr[i]);

            JsonSerializer.Serialize(writer, valueArr, options);
        }
    }
    [JsonConverter(typeof(EnumConverter<TestEnum>))]
    public class EnumConverter<T> : JsonConverter<T> where T : Enum
    {
        public override T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
            T o = (T)Activator.CreateInstance(typeof(T));

            return o;
        }

        public override void Write(Utf8JsonWriter writer, T theValue, JsonSerializerOptions options)
        {
            JsonSerializer.Serialize(writer, Convert.ToInt32(theValue), options);
        }
    }
`

The converters have to be set in the JsonSerializerOptions, you would use the above code is shown here:

`
            //Serialization options
            JsonSerializerOptions serializeOptions = new JsonSerializerOptions()
            {
                Converters = { new ValueTypeConverter<float>(),new ArrayConverter<float>() }
            };

            TestObject myObject = new TestObject();

            //Serialize the object
            string theSerializedObject = JsonSerializer.Serialize(myObject, options: serializeOptions);
            Debug.Log(theSerializedObject);

            //Then deserialize again
            TestObject myObjectDeserialized = JsonSerializer.Deserialize<TestObject>(theSerializedObject, serializeOptions);

            //Reserialize again to test consistency
            Debug.Log(JsonSerializer.Serialize(myObjectDeserialized, options: serializeOptions));
`

Of course, there is a downside to this method, it's that each time you want to serialize a value type you'll have to write a custom converter or at least set it up in JsonSerializationOptions. In our case, I wrote an ArrayConverter so that it takes advantage of the System.Text.Json ArrayConverter<A,B> without having to rewrite it.

**Comment by bdominguez** [comment]
@ponahoum are you sure that this works on IL2CPP? I've tried on 2020.3.0 LTS Windows 64bits IL2CPP with this result:

Repro project: [New Unity Project (1).zip](https://github.com/dotnet/runtime/files/6165988/New.Unity.Project.1.zip)


```
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.ExecutionEngineException: Attempting to call method 'System.Text.Json.JsonPropertyInfoNotNullable`4[[TestObject, Assembly-CSharp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null],[TestObject+TestEnum, Assembly-CSharp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null],[TestObject+TestEnum, Assembly-CSharp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null],[TestObject+TestEnum, Assembly-CSharp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]]::.ctor' for which no ahead of time (AOT) code was generated.
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00000] in <00000000000000000000000000000000>:0 
  at System.RuntimeType.CreateInstanceMono (System.Boolean nonPublic) [0x00000] in <00000000000000000000000000000000>:0 
  at System.RuntimeType.CreateInstanceDefaultCtor (System.Boolean publicOnly, System.Boolean skipCheckThis, System.Boolean fillCache, System.Threading.StackCrawlMark& stackMark) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Activator.CreateInstance (System.Type type, System.Boolean nonPublic) [0x00000] in <00000000000000000000000000000000>:0 
  at System.RuntimeType.CreateInstanceImpl (System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes, System.Threading.StackCrawlMark& stackMark) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Activator.CreateInstance (System.Type type, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Activator.CreateInstance (System.Type type, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.JsonClassInfo.CreateProperty (System.Type declaredPropertyType, System.Type runtimePropertyType, System.Type implementedPropertyType, System.Reflection.PropertyInfo propertyInfo, System.Type parentClassType, System.Text.Json.Serialization.JsonConverter converter, System.Text.Json.JsonSerializerOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.JsonClassInfo.AddProperty (System.Type propertyType, System.Reflection.PropertyInfo propertyInfo, System.Type classType, System.Text.Json.JsonSerializerOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.JsonClassInfo..ctor (System.Type type, System.Text.Json.JsonSerializerOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.JsonSerializerOptions.GetOrAddClass (System.Type classType) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.WriteStackFrame.Initialize (System.Type type, System.Text.Json.JsonSerializerOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.JsonSerializer.WriteCore (System.Text.Json.Utf8JsonWriter writer, System.Object value, System.Type type, System.Text.Json.JsonSerializerOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.JsonSerializer.WriteCore (System.Text.Json.PooledByteBufferWriter output, System.Object value, System.Type type, System.Text.Json.JsonSerializerOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.JsonSerializer.WriteCoreString (System.Object value, System.Type type, System.Text.Json.JsonSerializerOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.JsonSerializer.Serialize[TValue] (TValue value, System.Text.Json.JsonSerializerOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at NewBehaviourScript.Awake () [0x00000] in <00000000000000000000000000000000>:0 
   --- End of inner exception stack trace ---
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00000] in <00000000000000000000000000000000>:0 
  at System.RuntimeType.CreateInstanceMono (System.Boolean nonPublic) [0x00000] in <00000000000000000000000000000000>:0 
  at System.RuntimeType.CreateInstanceDefaultCtor (System.Boolean publicOnly, System.Boolean skipCheckThis, System.Boolean fillCache, System.Threading.StackCrawlMark& stackMark) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Activator.CreateInstance (System.Type type, System.Boolean nonPublic) [0x00000] in <00000000000000000000000000000000>:0 
  at System.RuntimeType.CreateInstanceImpl (System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes, System.Threading.StackCrawlMark& stackMark) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Activator.CreateInstance (System.Type type, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Activator.CreateInstance (System.Type type, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.JsonClassInfo.CreateProperty (System.Type declaredPropertyType, System.Type runtimePropertyType, System.Type implementedPropertyType, System.Reflection.PropertyInfo propertyInfo, System.Type parentClassType, System.Text.Json.Serialization.JsonConverter converter, System.Text.Json.JsonSerializerOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.JsonClassInfo.AddProperty (System.Type propertyType, System.Reflection.PropertyInfo propertyInfo, System.Type classType, System.Text.Json.JsonSerializerOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.JsonClassInfo..ctor (System.Type type, System.Text.Json.JsonSerializerOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.JsonSerializerOptions.GetOrAddClass (System.Type classType) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.WriteStackFrame.Initialize (System.Type type, System.Text.Json.JsonSerializerOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.JsonSerializer.WriteCore (System.Text.Json.Utf8JsonWriter writer, System.Object value, System.Type type, System.Text.Json.JsonSerializerOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.JsonSerializer.WriteCore (System.Text.Json.PooledByteBufferWriter output, System.Object value, System.Type type, System.Text.Json.JsonSerializerOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.JsonSerializer.WriteCoreString (System.Object value, System.Type type, System.Text.Json.JsonSerializerOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Json.JsonSerializer.Serialize[TValue] (TValue value, System.Text.Json.JsonSerializerOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at NewBehaviourScript.Awake () [0x00000] in <00000000000000000000000000000000>:0 
```

**Comment by ponahoum** [comment]
Yup, I was about to say it, but they fixed this issue in 5.0.0 ;)

**Comment by bdovaz** [comment]
@ponahoum what about structs like Color, DateTime, or Vector2? I have the following error despite of being added with "new ValueTypeConverter<Color>()", "new ValueTypeConverter<DateTime>()", "new ValueTypeConverter<Vector2>()"...

Vector2 example error:

```
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.ExecutionEngineException: Attempting to call method 'System.Text.Json.Serialization.Converters.ListOfTConverter`2[[System.Collections.Generic.List`1[[UnityEngine.Vector2, UnityEngine.CoreModule, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]], mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[UnityEngine.Vector2, UnityEngine.CoreModule, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]]::.ctor' for which no ahead of time (AOT) code was generated.
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Activator.CreateInstance (System.Type type, System.Boolean nonPublic) [0x00000] in <00000000000000000000000000000000>:0 
  at System.RuntimeType.CreateInstanceImpl (System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Object[] args, System.Globalization.CultureInfo culture, System.Object[] activationAttributes, System.Threading.StackCrawlMark& stackMark) [0x00000] in <00000000000000000000000000000000>:0 
 ```

**Comment by ponahoum** [comment]
Glad you solved your problem! I would also recommend using JsonConverterFactory to avoid having to declare  all the generic types in the JsonSerializerOptions, see [JsonConverterFactory documentation](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.serialization.jsonconverterfactory?view=net-5.0)

**Comment by bdovaz** [comment]
Can you give me an example of what could be achieved with that class? Do you mean that within that factory declare "by hand" all known types? int, decimal, double, long, float, ...

Because apart from those we would also have to include struct (DateTime, Vector2, Vector3), etc... It is impossible to anticipate all the built-in value types of .NET and Unity.

**Comment by ponahoum** [comment]
@bdovaz you can read more about it on this example : [Custom JsonConverter](https://www.thinktecture.com/en/asp-net/aspnet-core-3-0-custom-jsonconverter-for-the-new-system_text_json/)

Basically using this would allow you to declare a single converter in the JsonConverterOption per generic classes. However, you'll still need to reference the ListConverter<T> of you type T somewhere, but not specifically in the JsonConverterOption.

For example, if you have two enums "MyEnum", "MyOtherEnum", you'll have to declare ListConverter<MyEnum> and ListConverter<MyOtherEnum> somewhere in the code (just for the AOT compilation to include them in the build), but in the JsonConverterOption you'd just have to make a EnumConverterFactory. 

I use this option as I use a specific assembly for all the serialization operations of my app and this assembly, which declares the JsonSerializationOption, cannot access some of my classes (enum etc.), but this is a very specific use case though, and it doesn't solve the problem of having to manually declare all of the base converter somewhere at some point.
